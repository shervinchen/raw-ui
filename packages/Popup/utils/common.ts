type Prettify<T> = {
  [K in keyof T]: T[K];
} & {};

export type Side = 'top' | 'right' | 'bottom' | 'left';
export type Axis = 'x' | 'y';
export type Length = 'width' | 'height';
export type Strategy = 'absolute' | 'fixed';
export type Coords = { [key in Axis]: number };
export type Dimensions = { [key in Length]: number };
export type Rect = Prettify<Coords & Dimensions>;
export type SideObject = { [key in Side]: number };
export type ClientRectObject = Prettify<Rect & SideObject>;

export const round = Math.round;
export const createCoords = (v: number) => ({ x: v, y: v });
