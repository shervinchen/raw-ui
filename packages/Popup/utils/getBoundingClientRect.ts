import {
  ClientRectObject,
  Coords,
  createCoords,
  Dimensions,
  Rect,
  round,
} from './common';
import {
  isElement,
  isHTMLElement,
  getComputedStyle,
  getWindow,
  isWebKit,
  getFrameElement,
} from './dom';

const noOffsets = createCoords(0);

function getCssDimensions(element: Element): Dimensions & { $: boolean } {
  const css = getComputedStyle(element);
  let width = parseFloat(css.width) || 0;
  let height = parseFloat(css.height) || 0;
  const hasOffset = isHTMLElement(element);
  const offsetWidth = hasOffset ? element.offsetWidth : width;
  const offsetHeight = hasOffset ? element.offsetHeight : height;
  const shouldFallback =
    round(width) !== offsetWidth || round(height) !== offsetHeight;

  if (shouldFallback) {
    width = offsetWidth;
    height = offsetHeight;
  }

  return {
    width,
    height,
    $: shouldFallback,
  };
}

function getScale(element: Element): Coords {
  if (!isHTMLElement(element)) {
    return createCoords(1);
  }

  const rect = element.getBoundingClientRect();
  const { width, height, $ } = getCssDimensions(element);
  let x = ($ ? round(rect.width) : rect.width) / width;
  let y = ($ ? round(rect.height) : rect.height) / height;

  if (!x || !Number.isFinite(x)) {
    x = 1;
  }

  if (!y || !Number.isFinite(y)) {
    y = 1;
  }

  return {
    x,
    y,
  };
}

function getVisualOffsets(element: Element | undefined): Coords {
  const win = getWindow(element);

  if (!isWebKit() || !win.visualViewport) {
    return noOffsets;
  }

  return {
    x: win.visualViewport.offsetLeft,
    y: win.visualViewport.offsetTop,
  };
}

function shouldAddVisualOffsets(
  element: Element | undefined,
  isFixed = false,
  floatingOffsetParent?: Element | Window | undefined
): boolean {
  if (
    !floatingOffsetParent ||
    (isFixed && floatingOffsetParent !== getWindow(element))
  ) {
    return false;
  }

  return isFixed;
}

function rectToClientRect(rect: Rect): ClientRectObject {
  const { x, y, width, height } = rect;
  return {
    width,
    height,
    top: y,
    left: x,
    right: x + width,
    bottom: y + height,
    x,
    y,
  };
}

export function getBoundingClientRect(
  element: Element,
  includeScale = false,
  isFixedStrategy = false,
  offsetParent?: Element | Window
): ClientRectObject {
  const clientRect = element.getBoundingClientRect();

  let scale = createCoords(1);
  if (includeScale) {
    if (offsetParent) {
      if (isElement(offsetParent)) {
        scale = getScale(offsetParent);
      }
    } else {
      scale = getScale(element);
    }
  }

  const visualOffsets = shouldAddVisualOffsets(
    element,
    isFixedStrategy,
    offsetParent
  )
    ? getVisualOffsets(element)
    : createCoords(0);

  let x = (clientRect.left + visualOffsets.x) / scale.x;
  let y = (clientRect.top + visualOffsets.y) / scale.y;
  let width = clientRect.width / scale.x;
  let height = clientRect.height / scale.y;

  if (element) {
    const win = getWindow(element);
    const offsetWin =
      offsetParent && isElement(offsetParent)
        ? getWindow(offsetParent)
        : offsetParent;

    let currentWin = win;
    let currentIFrame = getFrameElement(currentWin);
    while (currentIFrame && offsetParent && offsetWin !== currentWin) {
      const iframeScale = getScale(currentIFrame);
      const iframeRect = currentIFrame.getBoundingClientRect();
      const css = getComputedStyle(currentIFrame);
      const left =
        iframeRect.left +
        (currentIFrame.clientLeft + parseFloat(css.paddingLeft)) *
          iframeScale.x;
      const top =
        iframeRect.top +
        (currentIFrame.clientTop + parseFloat(css.paddingTop)) * iframeScale.y;

      x *= iframeScale.x;
      y *= iframeScale.y;
      width *= iframeScale.x;
      height *= iframeScale.y;

      x += left;
      y += top;

      currentWin = getWindow(currentIFrame);
      currentIFrame = getFrameElement(currentWin);
    }
  }

  return rectToClientRect({ width, height, x, y });
}
