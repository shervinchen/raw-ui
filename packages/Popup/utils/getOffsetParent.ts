import {
  getContainingBlock,
  getDocumentElement,
  getParentNode,
  getWindow,
  isContainingBlock,
  isElement,
  isHTMLElement,
  isLastTraversableNode,
  isStaticPositioned,
  isTableElement,
  isTopLayer,
} from './dom';

function getTrueOffsetParent(element: Element): Element | null {
  if (
    !isHTMLElement(element) ||
    getComputedStyle(element).position === 'fixed'
  ) {
    return null;
  }

  let rawOffsetParent = element.offsetParent;

  if (getDocumentElement(element) === rawOffsetParent) {
    rawOffsetParent = rawOffsetParent.ownerDocument.body;
  }

  return rawOffsetParent;
}

export function getOffsetParent(element: Element): Element | Window {
  const win = getWindow(element);

  if (isTopLayer(element)) {
    return win;
  }

  if (!isHTMLElement(element)) {
    let svgOffsetParent = getParentNode(element);
    while (svgOffsetParent && !isLastTraversableNode(svgOffsetParent)) {
      if (isElement(svgOffsetParent) && !isStaticPositioned(svgOffsetParent)) {
        return svgOffsetParent;
      }
      svgOffsetParent = getParentNode(svgOffsetParent);
    }
    return win;
  }

  let offsetParent = getTrueOffsetParent(element);

  while (
    offsetParent &&
    isTableElement(offsetParent) &&
    isStaticPositioned(offsetParent)
  ) {
    offsetParent = getTrueOffsetParent(offsetParent);
  }

  if (
    offsetParent &&
    isLastTraversableNode(offsetParent) &&
    isStaticPositioned(offsetParent) &&
    !isContainingBlock(offsetParent)
  ) {
    return win;
  }

  return offsetParent || getContainingBlock(element) || win;
}
