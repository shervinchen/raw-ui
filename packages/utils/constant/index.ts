import { KeyCode } from './keyCode';

export { KeyCode };
