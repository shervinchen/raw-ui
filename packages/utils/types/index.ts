export type DeepPartial<T> = {
  [P in keyof T]?: T[P] extends Record<string, any> ? DeepPartial<T[P]> : T[P];
};

export type Merge<T, P> = P & Omit<T, keyof P>;
