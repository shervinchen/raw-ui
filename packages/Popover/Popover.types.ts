import { HTMLAttributes, MutableRefObject, ReactNode } from 'react';
import { PopupPlacement } from '../Popup/Popup.types';

interface BasePopoverProps {
  content: ReactNode;
  value?: boolean;
  defaultValue?: boolean;
  placement?: PopupPlacement;
  hideArrow?: boolean;
  disabled?: boolean;
  className?: string;
  getPopupContainer?: () => HTMLElement | null;
}

type NativePopoverProps = Omit<
  HTMLAttributes<HTMLDivElement>,
  keyof BasePopoverProps
>;

export type PopoverProps = BasePopoverProps & NativePopoverProps;

export interface PopoverArrowProps {
  placement: PopupPlacement;
  targetRef: MutableRefObject<HTMLElement | null>;
}

export interface PopoverArrowOffset {
  x: string;
  y: string;
}
