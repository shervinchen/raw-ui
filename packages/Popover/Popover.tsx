import React, { FC, PropsWithChildren, useRef } from 'react';
import classNames from 'classnames';
import { PopoverProps } from './Popover.types';
import { useClickAway, useControlled, useTransition } from '../utils/hooks';
import Popup from '../Popup';
import { computePopupPosition } from '../Popup/computePopup';
import { useTheme } from '../Theme';
import PopoverArrow from './PopoverArrow';

const Popover: FC<PropsWithChildren<PopoverProps>> = ({
  content,
  defaultValue = false,
  value,
  placement = 'top',
  hideArrow = false,
  disabled = false,
  className = '',
  getPopupContainer,
  onChange,
  children,
  ...restProps
}) => {
  const ref = useRef<HTMLDivElement>(null);
  const theme = useTheme();
  const [internalValue, setInternalValue] = useControlled<boolean>({
    defaultValue,
    value,
  });
  const { stage, shouldMount } = useTransition(internalValue, 50, 50);
  const classes = classNames('raw-popover', className);

  const clickHandler = () => {
    if (!disabled) {
      setInternalValue(!internalValue);
      onChange?.(!internalValue);
    }
  };

  useClickAway(
    ref,
    () => {
      if (internalValue) {
        setInternalValue(false);
      }
    },
    ['click']
  );

  return (
    <>
      <div ref={ref} className={classes} onClick={clickHandler} {...restProps}>
        {children}
      </div>
      <Popup
        name="popover"
        visible={shouldMount}
        targetRef={ref}
        getPopupPosition={() =>
          computePopupPosition(placement, ref, getPopupContainer)
        }
        getPopupContainer={getPopupContainer}
      >
        <div
          className="raw-popover-content"
          style={{
            opacity: stage === 'enter' ? 1 : 0,
          }}
          data-testid="popoverContent"
        >
          {!hideArrow && <PopoverArrow targetRef={ref} placement={placement} />}
          {content}
        </div>
      </Popup>
      <style jsx>{`
        .raw-popover {
          display: inline-flex;
        }
        .raw-popover-content {
          background-color: ${theme.palette.background};
          color: ${theme.palette.foreground};
          border-radius: 6px;
          border: 1px solid ${theme.palette.accents2};
          box-shadow: ${theme.tokens.shadow.sm};
          font-size: 14px;
          padding: 8px 12px;
          transition: opacity 0.05s ease;
        }
      `}</style>
    </>
  );
};

export default Popover;
