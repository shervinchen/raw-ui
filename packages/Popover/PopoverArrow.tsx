import { CSSProperties, FC } from 'react';
import css from 'styled-jsx/css';
import { PopoverArrowProps } from './Popover.types';
import PopupArrow from '../Popup/PopupArrow';
import { RawUITheme, useTheme } from '../Theme';
import { PopupPlacement } from '../Popup/Popup.types';

const getArrowBorderColor = (
  theme: RawUITheme
): { [key in PopupPlacement]: CSSProperties } => {
  return {
    top: {
      borderTop: 'transparent',
      borderLeft: 'transparent',
      borderBottom: theme.palette.neutral['200'],
      borderRight: theme.palette.neutral['200'],
    },
    topLeft: {
      borderTop: 'transparent',
      borderLeft: 'transparent',
      borderBottom: theme.palette.neutral['200'],
      borderRight: theme.palette.neutral['200'],
    },
    topRight: {
      borderTop: 'transparent',
      borderLeft: 'transparent',
      borderBottom: theme.palette.neutral['200'],
      borderRight: theme.palette.neutral['200'],
    },
    bottom: {
      borderTop: theme.palette.neutral['200'],
      borderLeft: theme.palette.neutral['200'],
      borderBottom: 'transparent',
      borderRight: 'transparent',
    },
    bottomLeft: {
      borderTop: theme.palette.neutral['200'],
      borderLeft: theme.palette.neutral['200'],
      borderBottom: 'transparent',
      borderRight: 'transparent',
    },
    bottomRight: {
      borderTop: theme.palette.neutral['200'],
      borderLeft: theme.palette.neutral['200'],
      borderBottom: 'transparent',
      borderRight: 'transparent',
    },
    left: {
      borderTop: theme.palette.neutral['200'],
      borderLeft: 'transparent',
      borderBottom: 'transparent',
      borderRight: theme.palette.neutral['200'],
    },
    leftTop: {
      borderTop: theme.palette.neutral['200'],
      borderLeft: 'transparent',
      borderBottom: 'transparent',
      borderRight: theme.palette.neutral['200'],
    },
    leftBottom: {
      borderTop: theme.palette.neutral['200'],
      borderLeft: 'transparent',
      borderBottom: 'transparent',
      borderRight: theme.palette.neutral['200'],
    },
    right: {
      borderTop: 'transparent',
      borderLeft: theme.palette.neutral['200'],
      borderBottom: theme.palette.neutral['200'],
      borderRight: 'transparent',
    },
    rightTop: {
      borderTop: 'transparent',
      borderLeft: theme.palette.neutral['200'],
      borderBottom: theme.palette.neutral['200'],
      borderRight: 'transparent',
    },
    rightBottom: {
      borderTop: 'transparent',
      borderLeft: theme.palette.neutral['200'],
      borderBottom: theme.palette.neutral['200'],
      borderRight: 'transparent',
    },
  };
};

const PopoverArrow: FC<PopoverArrowProps> = ({ targetRef, placement }) => {
  const theme = useTheme();
  const arrowBorderColor = getArrowBorderColor(theme);

  const { className, styles } = css.resolve`
    .raw-popup-arrow {
      background-color: ${theme.palette.background};
      border: 1px solid;
      border-top-color: ${arrowBorderColor[placement].borderTop};
      border-left-color: ${arrowBorderColor[placement].borderLeft};
      border-bottom-color: ${arrowBorderColor[placement].borderBottom};
      border-right-color: ${arrowBorderColor[placement].borderRight};
    }
  `;

  return (
    <>
      <PopupArrow
        targetRef={targetRef}
        placement={placement}
        className={className}
        withBorder
      />
      {styles}
    </>
  );
};

export default PopoverArrow;
