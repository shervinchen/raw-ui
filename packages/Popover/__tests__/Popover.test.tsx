import React, { useState } from 'react';
import { render, screen, waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import Popover from '..';
import { PopoverProps } from '../Popover.types';
import { KeyCode } from '../../utils/constant';

const user = userEvent.setup({ advanceTimers: jest.advanceTimersByTime });

describe('Popover', () => {
  test('should match the snapshot', () => {
    const { asFragment } = render(
      <Popover content="I am a popover">Click me</Popover>
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test('should show popover when default value is true', async () => {
    render(
      <Popover defaultValue content="I am a popover">
        Click me
      </Popover>
    );
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
  });

  test('should support custom class name', async () => {
    render(
      <Popover content="I am a popover" className="custom-popover">
        Click me
      </Popover>
    );
    await user.click(screen.getByTestId('popoverTarget'));
    expect(await screen.findByRole('dialog')).toHaveClass('custom-popover');
  });

  test('should show popover when click target', async () => {
    render(<Popover content="I am a popover">Click me</Popover>);
    await user.click(screen.getByTestId('popoverTarget'));
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
  });

  test('should switch whether or not popover is visible when click target', async () => {
    render(<Popover content="I am a popover">Click me</Popover>);
    const target = screen.getByTestId('popoverTarget');
    await user.click(target);
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
    await user.click(target);
    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  test('should hide popover when click outside', async () => {
    render(<Popover content="I am a popover">Click me</Popover>);
    await user.click(screen.getByTestId('popoverTarget'));
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
    await user.click(document.body);
    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  test('should support disabled popover', async () => {
    render(
      <Popover content="I am a popover" disabled>
        Click me
      </Popover>
    );
    await user.click(screen.getByTestId('popoverTarget'));
    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });

  test('should support controlled value', async () => {
    const onChange = jest.fn();

    const Component = (props: PopoverProps) => {
      const [value, setValue] = useState(false);

      return (
        <Popover
          content={props.content}
          value={value}
          onChange={(nextValue) => {
            setValue(nextValue);
            props.onChange?.(nextValue);
          }}
        >
          Click me
        </Popover>
      );
    };
    render(
      <Component content="I am a controlled popover" onChange={onChange} />
    );
    await user.click(screen.getByTestId('popoverTarget'));
    expect(await screen.findByRole('dialog')).toBeInTheDocument();
    expect(onChange).toHaveBeenCalledTimes(1);
  });

  test('should support set parent container element', () => {
    render(
      <div
        id="parentElement"
        style={{
          position: 'relative',
          overflowY: 'auto',
          width: '400px',
          height: '200px',
        }}
      >
        <Popover
          content="I am a popover"
          getPopupContainer={() => document.querySelector('#parentElement')}
        >
          Click me
        </Popover>
      </div>
    );
  });

  test('should close popover when press the Escape key', async () => {
    render(
      <Popover defaultValue content="I am a popover">
        Click me
      </Popover>
    );
    expect(screen.getByRole('dialog')).toBeInTheDocument();
    await user.keyboard(`[${KeyCode.Escape}]`);
    await waitFor(() => {
      expect(screen.queryByRole('dialog')).not.toBeInTheDocument();
    });
  });
});
