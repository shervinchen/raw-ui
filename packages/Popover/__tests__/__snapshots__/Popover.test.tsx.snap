// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover should match the snapshot 1`] = `
<DocumentFragment>
  <div
    class="jsx-3851428059 raw-popover"
    data-testid="popoverTarget"
  >
    Click me
  </div>
</DocumentFragment>
`;
